gen/bind_generator.jl

```diff
@@ -0,0 +1,147 @@
+push!(bind_exceptions, :h5e_set_auto => :H5Eset_auto2)
+push!(bind_exceptions, :h5g_create => :H5Gcreate2)
+push!(bind_exceptions, :h5g_open => :H5Gopen2)
+push!(bind_exceptions, :h5o_get_info => :H5Oget_info1)
+push!(bind_exceptions, :h5r_get_obj_type => :H5Rget_obj_type2)
+push!(bind_exceptions, :h5t_array_create => :H5Tarray_create2)
+push!(bind_exceptions, :h5t_commit => :H5Tcommit2)
+push!(bind_exceptions, :h5t_get_array_dims => :H5Tget_array_dims2)
+push!(bind_exceptions, :h5t_open => :H5Topen2)
+
+# Distinguishes 32-bit vs 64-bit handle arguments
+push!(bind_exceptions, :h5p_get_fapl_mpio32 => :H5Pget_fapl_mpio)
+push!(bind_exceptions, :h5p_get_fapl_mpio64 => :H5Pget_fapl_mpio)
+push!(bind_exceptions, :h5p_set_fapl_mpio32 => :H5Pset_fapl_mpio)
+push!(bind_exceptions, :h5p_set_fapl_mpio64 => :H5Pset_fapl_mpio)
+
+"""
+    @bind h5_function(arg1::Arg1Type, ...)::ReturnType ErrorStringOrExpression
+
+A binding generator for translating `@ccall`-like declarations of HDF5 library functions
+to error-checked `ccall` expressions.
+
+The provided function name is used to define the Julia function. The corresponding C
+function name is auto-generated by uppercasing the first few letters (up to the first `_`),
+and the first `_` is removed. Explicit name mappings can be made by inserting a
+`:jlname => :h5name` pair into the `bind_exceptions` dictionary.
+
+The `ErrorStringOrExpression` can be either a string literal or an arbitrary expression.
+If a `String`, the string is used as the message in an `error()` call, otherwise the
+expression is used as-is. Note that the expression may refer to any arguments by name.
+
+The declared return type in the function-like signature must be the return type of the C
+function, and the Julia return type is inferred as one of the following possibilities:
+
+1. If `ReturnType === :Herr`, the Julia function returns `nothing` and the C return is
+   used only in error checking.
+
+2. If `ReturnType === :Htri`, the Julia function returns a boolean indicating whether
+   the return value of the C function was zero (`false`) or positive (`true`).
+
+3. Otherwise, the C function return value is returned from the Julia function.
+
+Furthermore, the C return value is interpreted to automatically generate error checks
+in most cases.
+
+1. If `ReturnType === :Herr` or `ReturnType === :Htri`, an error is raised when the return
+   value is negative.
+
+2. If `ReturnType === :Haddr` or `ReturnType === :Hsize`, an error is raised when the
+   return value is equivalent to `-1 % Haddr` and `-1 % Hsize`, respectively.
+
+3. For all other return types, it is assumed a negative value indicates error.
+
+It is assumed that the HDF library names are given in global constants named `libhdf5`
+and `libhdf5_hl`. The former is used for all `ccall`s, except if the C library name begins
+with "H5DO" or "H5TB" then the latter library is used.
+"""
+macro bind(sig::Expr, err::Union{String,Expr})
```

Comment:
Come to think of it, wouldn't have been  better if the `@bind` generator converted the `C` function name to the  Julia name as opposed to the opposite process ?  This is in some sense similar to the `@ccall` macro.

Reply:
I guess there's the issue of the exception list, which complicates that a bit. I assume that's probably why you took this approach instead.

Reply:
Yeah, I'd originally done it the other way around, but the MPIO wrappers which all share the same C call but have unique Julia names made that complicated since the Dict key wasn't unique (without having to do something more complicated like using the full type signature as the key).